lib/discourse/sso.ex

```diff
@@ -0,0 +1,215 @@
+defmodule Discourse.SSO do
+  @moduledoc """
+  kDiscourse SSO.
+
+  Reference: https://meta.discourse.org/t/official-single-sign-on-for-discourse-sso/13045
+  """
+  require Logger
+
+  @string_fields ~w(
+    add_groups
+    avatar_url
+    bio
+    card_background_url
+    email
+    external_id
+    groups
+    locale
+    name
+    nonce
+    profile_background_url
+    remove_groups
+    return_sso_url
+    title
+    username
+    website
+  )a
+
+  @boolean_fields ~w(
+    admin
+    avatar_force_update
+    locale_force_update
+    moderator
+    require_activation
+    suppress_welcome_message
+  )
+
+  @fields @string_fields ++ @boolean_fields
+
+  @doc ~S"""
+  Verify the passed sso nonce and signature.
+
+  The `secret` set for Discourse is taken from the application configurations,
+  but can also be supplied as option.
+
+  The passed `secret` takes priority over the configured one.
+
+  ## Example
+
+  ```
+  iex> SSO.validate("bm9uY2U9Y2I2ODI1MWVlZmI1MjExZTU4YzAwZmYxMzk1ZjBjMGI=\n", "2828aa29899722b35a2f191d34ef9b3ce695e0e6eeec47deb46d588d70c7cb56")
+  {:ok, "cb68251eefb5211e58c00ff1395f0c0b"}
+  ```
+  """
+  @spec validate(String.t(), String.t(), Keyword.t()) ::
+          {:ok, nonce :: String.t()} | {:error, atom}
+  def validate(sso, sig, opts \\ []) do
+    with true <- sig(sso, opts) == sig,
+         {:ok, decoded} <- Base.decode64(String.trim(sso, "\n"), padding: true),
+         %{"nonce" => nonce} <- URI.decode_query(decoded) do
+      {:ok, nonce}
+    else
+      error = {:error, _} -> error
+      false -> {:error, :invalid_signature}
+      _ -> {:error, :invalid_payload}
+    end
+  end
+
+  @doc ~S"""
+  Creates a signed url to redirect users to.
+
+  The following user data is required:
+
+    - `id`, the id of the user in your system.
+    - `email`, the email of the user. (Assumed to be verified.)
+    - `nonce`, the nonce given at the start of the request.
```

Comment:
What's a nonce?

Reply:
A number or bit string used only once.

It's part of the Discourse SSO flow, you get it from `validate/2` or `validate/3`. (`{:ok, nonce :: String.t}`)